### PR TITLE
make sure events are only ignored if they are Ignoreable

### LINKED ISSUE #77: Version 1.6.0 breaks PR builds with Bitbucket branch source plugin
I have tested Bitbucket branch source plugin in v2.2.7 and v2.2.5 and PRs are broken in both versions
My best guess it Jenkins getting unexpected JSON response but I am not sure.
Debugging on the side of Jenkins didn't prove anything useful.

I will try and set up a bitbucket/jenkins debug session and see if I catch anything useful.

### PR DESCRIPTION
**Pull request by PixelJonas**
In order to fix #77 I added two new Interfaces

`Event ` -> Functional Interface to identify our custom events @casz can you please have a look if I tagges all events correctly?

`Ignorable` -> Interface to indicate this event can be ignored by commiters/branches

I'm not quite certain on how to ignore PR-Events... Should we ignore PRs from "ignored commiters" or on "ignored branches" ?

### CODE REVIEW

File: src/test/java/nl/topicus/bitbucket/api/PushEventServiceTest.java [hunk 1]
```diff
@@ -1,9 +1,7 @@
 package nl.topicus.bitbucket.api;
 
 import com.google.common.collect.ImmutableList;
-import nl.topicus.bitbucket.events.BitbucketPushChange;
-import nl.topicus.bitbucket.events.BitbucketPushDetail;
-import nl.topicus.bitbucket.events.BitbucketPushEvent;
+import nl.topicus.bitbucket.events.*;
```
Comment (casz) [inline src/test/java/nl/topicus/bitbucket/api/PushEventServiceTest.java]: Explicit imports, please 🙏

File: src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java [hunk 2]
```diff
@@ -1,9 +1,9 @@
 package nl.topicus.bitbucket.api;
 
 import com.atlassian.bitbucket.ServiceException;
+import com.atlassian.bitbucket.build.BuildStatusSetEvent;
 import com.atlassian.bitbucket.event.branch.BranchCreatedEvent;
 import com.atlassian.bitbucket.event.branch.BranchDeletedEvent;
-import com.atlassian.bitbucket.build.BuildStatusSetEvent;
 import com.atlassian.bitbucket.event.pull.*;
```
Comment (casz) [inline src/main/java/nl/topicus/bitbucket/api/PullRequestListener.java]: While your at it, can we have explicit imports 👍

**Comment by casz** [comment]
I do not think pull request should be ignored, that would defeat the purpose.

Disregard my request for explicit, imports I see plenty more implicit imports.

**Comment by PixelJonas** [comment]
seems like my codestyle settings got mixed up. Whole project should now use explicit imports ;)

**Comment by casz** [comment]
Thanks 👍

**Comment by casz** [comment]
@PixelJonas Thanks for your contribution ❤️

**Comment by casz** [comment]
oh noes @PixelJonas 
```
2017-11-28 10:39:24,698 ERROR [threadpool:thread-4]  c.a.s.i.c.DefaultTransferableStateManager Error while processing asynchronous task
java.lang.NullPointerException: null
	at nl.topicus.bitbucket.api.PushEventService.ignoredByCommiters(PushEventService.java:34)
	at nl.topicus.bitbucket.api.PushEventService.isIgnoredEvent(PushEventService.java:30)
	at nl.topicus.bitbucket.api.PushEventService.isValidEvent(PushEventService.java:26)
	at nl.topicus.bitbucket.api.PullRequestListener.sendEvents(PullRequestListener.java:313)
	at nl.topicus.bitbucket.api.PullRequestListener.lambda$onRepositoryRefsChanged$0(PullRequestListener.java:202)
	at com.atlassian.stash.internal.concurrent.DefaultTransferableStateManager$StateTransferringRunnable.run(DefaultTransferableStateManager.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(Unknown Source)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
	... 1 frame trimmed
```

Looking into it :|

File: src/main/java/nl/topicus/bitbucket/events/BitbucketPushEvent.java [hunk 3]
```diff
@@ -27,42 +27,56 @@
...
         this.push = push;
     }
+
+    @Override
+    public Optional<String> getUsername() {
+        return Optional.ofNullable(this.getActor() == null ? null : this.getActor().getUsername());
+    }
+
+    @Override
+    public List<String> getBranches() {
```
Comment (casz) [inline src/main/java/nl/topicus/bitbucket/events/BitbucketPushEvent.java]: Rats this causes #80 is there no end to this unfortunate luck @PixelJonas

File: src/main/java/nl/topicus/bitbucket/events/BitbucketPushEvent.java [hunk 4]
```diff
@@ -27,42 +27,56 @@
...
+        if (this.getPush() == null || this.getPush().getChanges() == null) {
+            return new ArrayList<>();
+        }
+        return this.getPush().getChanges()
+                .stream()
+                .map(bitbucketPushChange -> bitbucketPushChange.getNew().getName())
```
Comment (casz) [inline src/main/java/nl/topicus/bitbucket/events/BitbucketPushEvent.java]: Which causes a null pointer on getNew 😭

**Pull request merged**